include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+{
+    struct otTcpActiveSocket *mNext;                  ///< A pointer to the next active TCP socket (internal use only)
+    void *mContext;                                   ///< A pointer to application-specific context
+    otTcpConnectDone mConnectDoneCallback;            ///< "Connect done" callback function
+    otTcpSendDone mSendDoneCallback;                  ///< "Send done" callback function
+    otTcpSendReady mSendReadyCallback;                ///< "Send ready" callback function
+    otTcpReceiveAvailable mReceiveAvailableCallback;  ///< "Receive available" callback function
+    otTcpConnectionLost mConnectionLostCallback;      ///< "Connection lost" callback function
+
+    /* Other implementation-defined fields go here. */
+} otTcpActiveSocket;
+
+/**
+ * This structure contains arguments to the @p otTcpActiveOpen function.
+ */
+typedef struct otTcpActiveOpenArgs
+{
+    otTcpConnectDone mConnectDoneCallback;            ///< "Connect done" callback function
+    otTcpSendDone mSendDoneCallback;                  ///< "Send done" callback function
+    otTcpBytesAcked mBytesAckedCallback;              ///< "Bytes acked" callback
+    otTcpSendReady mSendReadyCallback;                ///< "Send ready" callback function
+    otTcpReceiveAvailable mReceiveAvailableCallback;  ///< "Receive available" callback function
+    otTcpConnectionLost mConnectionLostCallback;      ///< "Connection lost" callback function
+    void *mContext;                                   ///< Pointer to application-specific context
+    void *mReceiveBuffer;                             ///< Pointer to memory provided to the system for the TCP receive buffer
+    size_t mReceiveBufferSize;                        ///< Size of memory provided to the system for the TCP receive buffer
+} otTcpActiveOpenArgs;
+
+/**
+ * Opens an active TCP socket.
+ *
+ * @param[in]  aInstance  A pointer to an OpenThread instance.
+ * @param[in]  aSocket    A pointer to an active TCP socket structure.
+ * @param[in]  aArgs      A pointer to a structure of arguments.
+ * @param[in]  aContext   A pointer to application-specific context.
+ *
+ * @retval OT_ERROR_NONE    Successfully opened the socket.
+ * @retval OT_ERROR_FAILED  Failed to open the socket.
+ */
+otError otTcpActiveOpen(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpActiveOpenArgs *aArgs);
+
+/**
+ * Sets the "connect done" callback on an active TCP socket.
+ *
+ * @param[in]  aInstance             A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket               A pointer to the active TCP socket structure whose callback to set.
+ * @param[in]  aConnectDoneCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetConnectDoneCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpConnectDone aConnectDoneCallback);
```

Comment:
Callback can be configured by `otTcpActiveOpen` already. 
Should we remove callback setting APIs?

_We can add them when someone find it useful._

Reply:
I'm open to this. But, in the context of libraries (e.g., one that implements a circular buffer), the library may want to set the callback functions differently. We could also just have them set the fields inside `otTcpActiveSocket` or `otTcpPassiveSocket` though.